algorithms/src/snark/marlin/marlin.rs

```diff
@@ -559,33 +612,101 @@ where
+        let proof = Proof::<E>::new(batch_sizes, total_batch_size, commitments, evaluations, prover_third_message, pc_proof)?;
         assert_eq!(proof.pc_proof.is_hiding(), MM::ZK);
 
+        // Collect verification key and public inputs to verify_batch
+        let mut vk_to_inputs = BTreeMap::new();
+        for pk in keys_to_constraints.keys() {
+            vk_to_inputs.insert(&pk.circuit_verifying_key, public_inputs.get(&pk.circuit.hash).unwrap().as_slice());
+        }
+
         #[cfg(debug_assertions)]
-        if !Self::verify_batch(fs_parameters, &circuit_proving_key.circuit_verifying_key, &public_input, &proof)? {
+        if !Self::verify_batch(fs_parameters, &vk_to_inputs, &proof)? {
             println!("Invalid proof")
         }
         end_timer!(prover_time);
 
         Ok(proof)
     }
 
-    fn verify_batch_prepared<B: Borrow<Self::VerifierInput>>(
+    fn verify_batch_prepared<'a, B: Borrow<Self::VerifierInput>>(
         fs_parameters: &Self::FSParameters,
-        prepared_verifying_key: &<Self::VerifyingKey as Prepare>::Prepared,
-        public_inputs: &[B],
+        keys_to_inputs: &BTreeMap<<Self::VerifyingKey as PrepareOrd>::Prepared, &[B]>,
         proof: &Self::Proof,
     ) -> Result<bool, SNARKError> {
-        let circuit_verifying_key = &prepared_verifying_key.orig_vk;
-        if public_inputs.is_empty() {
+
+        if keys_to_inputs.len() == 0 {
             return Err(SNARKError::EmptyBatch);
         }
 
-        if public_inputs.len() != proof.batch_size()? {
-            return Err(SNARKError::BatchSizeMismatch);
+        let batch_sizes = proof.batch_sizes()?;
+        for (vk, public_inputs_i) in keys_to_inputs {
+            if public_inputs_i.is_empty() {
+                return Err(SNARKError::EmptyBatch);
+            }
+
+            if public_inputs_i.len() != proof.batch_sizes()?[&vk.orig_vk.hash] {
+                return Err(SNARKError::BatchSizeMismatch);
+            }
         }
 
+        // collect values into structures for our calculations
+        let mut max_num_constraints = 0;
+        let mut max_num_non_zero = 0;
+        let mut public_inputs = BTreeMap::new();
+        let mut padded_public_vec = Vec::with_capacity(keys_to_inputs.len());
+        let mut padded_public_inputs = BTreeMap::new();
+        let mut input_domains = BTreeMap::new();
+        let mut circuit_infos = BTreeMap::new();
+        let mut vks = Vec::with_capacity(keys_to_inputs.len());
+        for (i, (vk, public_inputs_i)) in keys_to_inputs.iter().enumerate() {
+            vks.push(&vk.orig_vk.verifier_key);
+
+            max_num_constraints = max_num_constraints.max(vk.orig_vk.circuit_info.num_constraints);
+            max_num_non_zero = max_num_non_zero.max(vk.orig_vk.circuit_info.num_non_zero_a);
+            max_num_non_zero = max_num_non_zero.max(vk.orig_vk.circuit_info.num_non_zero_b);
+            max_num_non_zero = max_num_non_zero.max(vk.orig_vk.circuit_info.num_non_zero_c);
+
```

Comment:
It might be worth extracting this into a common function between the prover and verifier